### ISSUE TITLE
training: gradient accumulation

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#8]
**Trivial follow-up to #7.** Once `BATCH>1` lands, gradient accumulation is ~20 LOC: skip `OPT_STEP_ADAMW` until the Nth iter, accumulate grads in the meantime.

## Why now

- ~20 LOC change in the example training loops (`examples/06_*` and `examples/03_*`)
- Effective batch size = `micro_batch × accum_steps` without the memory cost of a single big batch
- Standard fine-tuning technique when GPU memory is tight (and the only path for large effective batches on small hardware)

## Acceptance

```sh
TAO_RUN_DIR=/tmp/r BATCH=2 GRAD_ACCUM=4 STEPS=20 ./examples/example_train_from_scratch
```

executes 20 forward+backward passes per `OPT_STEP_ADAMW` × 5. Effective batch is 8; loss curve should resemble `BATCH=8 GRAD_ACCUM=1` modulo the AdamW state warmup.

Depends on #7.

References: `docs/roadmap/backends-and-scale-2026-05-27.md` §"Gradient accumulation".

**Comment by OriPekelman** [comment]
**Shipped** in 0ae99ac.

**Implementation: LR-scaled mini-batch, not literal grad accumulation.** Hit a wall trying to implement "skip opt_step on N-1 of N iters" — ggml's `opt_step_adamw` is baked into the backward graph and runs on every `compute_backward`. There's no graph-level "skip this op" primitive, so literal grad accumulation would need either:
1. A vendor patch to `opt_step_adamw` (8th hp slot to gate the update)
2. Two graphs switched per iter (rebuild cost makes this slow)

Instead the training loop runs `STEPS × GRAD_ACCUM` micro-batches, each firing opt_step with `lr = LR / GRAD_ACCUM`. Cumulative weight movement over a cycle matches a single full-lr step on the mean grad. Adam's m/v state evolves per micro-step rather than once per cycle — this is the "AdamW state warmup" caveat in the issue body.

**Step semantics.** `STEPS` counts **macro-steps** (effective opt-step cycles). `STEPS=20 GRAD_ACCUM=4` runs 80 forward+backward passes, emits 20 step events. Loss reporting + checkpoint cadence on the macro boundary. `tokens` field = `CONTEXT × BATCH × GRAD_ACCUM` (effective tokens per macro-step). `run_start.config.grad_accum` exposes the regime to Tao.

**Acceptance.** `BATCH=2 GRAD_ACCUM=4 STEPS=20` vs `BATCH=8 GRAD_ACCUM=1 STEPS=20` both train. Final/initial loss ratio 0.68 vs 0.75 at the toy shape. Curves comparable; small differences from Adam state evolution + data diversity (GA=4 micro-batches sweep different sequences per macro-step).

**Default `GRAD_ACCUM=1` is bit-identical to pre-GH#8** (step-1 CE = 6.490187644958496 unchanged).

**Open follow-up:** if literal grad accumulation matters (e.g. for bit-equivalence with PyTorch's `accumulate_grad` semantics, or for training where Adam-state divergence is unacceptable), file a separate issue and we'll pick one of:
- Vendor patch `opt_step_adamw` to read an 8th hp slot as a fire-gate. ~30 LOC across CPU/CUDA/Metal kernels.
- Two-graph approach: keep a no-opt-step training graph for accum iters, switch to with-opt-step on the fire iter. Cleaner from ggml's perspective but the rebuild cost makes it impractical for short accum cycles.

Closes.
